tools/lan_choose_character_kart/lan_choose_character_kart.py

```diff
@@ -0,0 +1,299 @@
+        run_subprocess(
+            [devkit_tools.LDPATH, asmfileobj_path, '-T', symbolsld_path, '-o', asmfileldobj_path])
+
+        # Extract only the machine code from the ELF file
+        devkit_tools.objcopy(asmfileldobj_path, '-O', 'binary', '-S')
+
+        with open(asmfileldobj_path, 'rb') as obj_file:
+            machinecode = obj_file.read()
+            assert len(
+                machinecode) != 0, f'Error with {cb.curasmfile}, did not compile to anything!'
+            finaldata_output += machinecode
+
+    return finaldata_output
+
+
+DOL_TEXTOFFSET: int = 0x800030c0
+
+
+def patch_maindol(region: str, dol_path: str, temp_dir: tempfile.TemporaryDirectory,
+                  asm_dir: str) -> None:
+    patchconfig = (
+        ("SceneLanEntry__SceneLanEntry_after_Scene__Scene_call", patchcmd_asmfile,
+         'patchlogoapp.s'),
+        (
+            "LANEntry__waitEntry_before_LANEntry__endReceipt",
+            patchcmd_asm,
+            ("li r0, 0x3", "stw r0, 0x0(r25)")  # Prevent endReceipt from being called
+        ),
+        ("LANEntry__calc_cmpwi_3_bge", patchcmd_bge, "LANEntry__calc_LANEntry__setRaceInfo_call"),
+        ("LANEntry__calc_LANEntry__setRaceInfo_call", patchcmd_asmfile,
+         "lanentrycalc_setraceinfo.s"),
+        ("LANEntry_alloc_size", patchcmd_asm,
+         "li r3, 0x33a"),  # Adjust alloc size to account for new fields used by character/kart menu
+        ("LANEntry__draw_LANEntry_blo_draw", patchcmd_bl, "LANEntryDrawBeforeSecondDrawResolved"),
+        ("LANEntry__calcAnm_lwz_JUTFader_vtable", patchcmd_asmfile, "lanentrycalcanm_timerup.s"),
+        ("NetGateApp__NetGateApp_blr", patchcmd_b, "NetGateAppAfterCtResolved"),
+        ("NetGateApp_alloc_size", patchcmd_asm,
+         "li r3, 0x34+0x4"),  # Adjust alloc size to account for pointer to arrow texture
+        ("LANEntry__LANEntry_epilogue", patchcmd_bl, "SetRandSeedAfterLanEntryCtResolved"),
+    )
+
+    with open(dol_path, 'rb') as dol:
+        dol = bytearray(dol.read())
+
+    for sym, cmd, arg in patchconfig:
+        adr = symbols.symbols[region][sym]
+        cmdret = cmd(adr, arg, region, temp_dir, asm_dir)
+        adr_resolved = adr - DOL_TEXTOFFSET
+        dol[adr_resolved:adr_resolved + len(cmdret)] = cmdret
+
+    with open(dol_path, 'wb') as patcheddol:
+        patcheddol.write(dol)
+
+
+def patchcmd_asm(_adr: int, instructions: str | tuple[str], _region: str,
+                 temp_dir: tempfile.TemporaryDirectory, asm_dir: str) -> bytes:
+    asmfile_path = os.path.join(temp_dir, 'temp.s')
+    asmfileobj_path = os.path.join(temp_dir, 'temp.o')
+
+    if isinstance(instructions, str):
+        instructions = (instructions, )
+    assert isinstance(instructions, tuple), f'instructions variable {instructions} is not a tuple!'
+
+    with open(asmfile_path, 'w', encoding='utf-8') as tempasm:
+        tempasm.write('\n'.join(instructions))
```

Comment:
It'd be nice to add a trailing `\n` character to content the assembler, as otherwise it throws this warning:

```
/tmp/mkddextyhh9e8g0/temp.s: Assembler messages:
/tmp/mkddextyhh9e8g0/temp.s: Warning: end of file not at end of a line; newline inserted
/tmp/mkddextyhh9e8g0/temp.s: Assembler messages:
/tmp/mkddextyhh9e8g0/temp.s: Warning: end of file not at end of a line; newline inserted
/tmp/mkddextyhh9e8g0/temp.s: Assembler messages:
/tmp/mkddextyhh9e8g0/temp.s: Warning: end of file not at end of a line; newline inserted
```

```suggestion
    with open(asmfile_path, 'w', encoding='utf-8') as tempasm:
        tempasm.write('\n'.join(instructions) + '\n')
```